jetty-io/src/main/java/org/eclipse/jetty/io/AdapterMemoryPool.java

```diff
@@ -0,0 +1,42 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.nio.ByteBuffer;
+import java.util.function.Consumer;
+
+public class AdapterMemoryPool implements MemoryPool<RetainableByteBuffer>
```

Comment:
Javadoc needed.   Not sure why the name is "Adapter"?

Why not `RetainableMemoryPool`?    Hmmm but that is confusing with RBBP??? OK not seeing what this class is for?

Reply:
`AdapterMemoryPool` adapts the old `ByteBufferPool` to the new `RetainableByteBufferPool` abstracted out as `MemoryPool<RetainableByteBuffer>`. This allows the input side to use `MemoryPool<RetainableByteBuffer>` only while still being able to work with `ByteBufferPool` implementations.

The reason for this mechanism is backward compatibility: there are components that used to expect a `ByteBufferPool` that now need a `MemoryPool<RetainableByteBuffer>`. Special care is going to be needed to figure out exactly what breaking changes we're okay with and what needs to be kept compatible.

Reply:
So for backwards compatibility we have the two cases to consider:
1. Somebody has written a component expecting to be able to find a ByteBufferPool as a bean on the server, but we have only added a RetainableByteBuffer.
2. We have a component that now expects to find a RetainableByteBuffer on the server, but the deployer has only added a ByteBufferPool

So this adaptor is useful for the later case, but do we have something that will work for the former case?


Perhaps there should be some static methods somewhere that capture both search algorithms:

1. look for a ByteBufferPool, if found use it, else look for a RetainableButeBufferPool, if found adapt it??? else create a new ByteBufferPool
2. look for a RetainableByteBufferPool, if found use it, else look for a ByteBufferPool, if found then adapt it (and emit a warning or at least an info), else create a new RetainableByteBufferPool.

Reply:
Case 1 (_Somebody has written a component expecting to be able to find a ByteBufferPool as a bean on the server, but we have only added a RetainableByteBuffer._) cannot happen as we still rely on `ByteBufferPool` for output buffers, so there should always be one available.

Maybe we should log when we cannot find a `RetainableByteBufferPool` and have to use an adapter, but @sbordet usually does not like such warnings.

Reply:
Wouldn't it be better to have the RBBP wrapped to look like a plain BBP?    Although that will just increase contention... and if input/output buffers sizes are different, then there is little overlap.     OK.